0025-nat-traversal.adoc

```diff
@@ -0,0 +1,174 @@
+
+=== STUN/Hole Punching
+
+Bob sends UDP messages to a third peer known as a STUN server.
+The STUN server will return which ports and public IP Bobs's UDP messages originated from.
+Bob has now discovered public IP and ports which peers can use to dial him.
+Bob can provide this knowledge to Alice, allowing Alice to dial him.
+
+STUN cannot be used to traverse NAT configurations that randomly assign a port and/or IP for new connections.
+For example, a symmetric NAT will randomly generate a new port and/or public IP for each outbound connection made by Alice.
+This prevents Bob from determining or predicting the external IP and ports assigned to her by the NAT.
+
+- + Does not require Bob to configure his NAT
+
+- + Alice and Bob can communicate through a direct connection
+
+- - Cannot traverse symmetric, cone and restricted cone type NATs
+
+- - Requires a third peer to start a connection between Alice and Bob
+
+- - Requires the third peer to know its public IP and ports (either through another STUN/TURN or port forwarding)
+
+- - Requires peers to know peers that are STUN servers
+
+- - Leads to some degree of centralisation
+
+=== TURN
+
+When Bob isn’t able to determine or predict his public facing IP address and port, Bob can dial out to a relay peer, which will keep a long-lived connection open.
+Other peers will be able to dial through the relay peer, which will forward traffic to Bob.
+
+- + Can be used to traverse all commonly know NAT configurations
+
+- + Does not require the Bob to configure his NAT
+
+- - Requires a third peer to serve as a relay for communications between Alice and Bob
+
+- - Requires the third peer to know its inbound ports (either through another STUN/TURN or port forwarding)
+
+- - Requires a third peer to serve as a relay for communications between Alice and Bob
+
+- - Requires peers to know peers that are TURN servers
+
+- - Leads to some degree of centralisation
+
+== Recommendation
+
+Assumption: A significant portion of comit nodes will have forwarded ports.
+
+We should implement the following NAT traversal methods in the order listed below.
+After each implementation we should evaluate whether it is necessary to implement the subsequent NAT traversal methods.
+
+1. TURN (essential for maximum NAT coverage)
+2. STUN (optional to reduce load and reliance on TURN servers)
+3. UPnP (optional to reduce load and reliance on TURN servers)
+
+TURN is the first prioirty as it is capable of traversing most commonly known NAT configurations.
+In order for TURN to work, some port forwarded peers will have to operate as TURN servers to allow NAT restricted peers to communicate.
+
+The libp2p specification defines a circuit relay module based off the TURN protocol.
+This protocol allows NAT restricted nodes to "proxy" through other p2p nodes that volunteer to act as p2p-circuit relays.
+These nodes will announce to the network that they are accepting relay connections using libp2p's peer routing.
```

Comment:
How do we get started initially? i.e. how does a node behind NAT know about these relay nodes?

Reply:
The simplest solution would be to allow the use to specify a list of relay nodes through a config file. We could provide a default list of nodes that we approve of/ are run by us/ have a history of serving as reliable relays.

To further improve on this solution you could ask your "friends" or peers that you trust for a list of relay nodes that they trust. At this stage we are entering the realm of decentralised trust systems. The general problem is: "How do I know which peers are trustworthy providers of information?".  The information could be a list of reliable relay nodes or orders from an orderbook. I don't think this an easy problem to address and I don't think a potential solution to this problem should be a part of this spike.